--- a/dwm.c	2012-02-12 10:56:30.098137240 +0200
+++ b/dwm.c	2012-02-12 10:56:30.100137240 +0200
@@ -265,6 +265,9 @@ static void bstack(Monitor *m);
 static Client *prevtiled(Client *c);
 static void pushup(const Arg *arg);
 static void pushdown(const Arg *arg);
+static void cycle(const Arg *arg);
+static int shifttag(int dist);
+static void tagcycle(const Arg *arg);
 
 /* variables */
 static const char broken[] = "broken";
@@ -2277,6 +2280,42 @@ pushdown(const Arg *arg) {
 }
 
 int
+shifttag(int dist) {
+	int i, curtags;
+	int seltag = 0;
+	int numtags = LENGTH(tags);
+
+	curtags = selmon->tagset[selmon->seltags];
+	for(i = 0; i < LENGTH(tags); i++) {
+		if((curtags & (1 << i)) != 0) {
+			seltag = i;
+			break;
+		}
+	}
+
+	seltag += dist;
+	if(seltag < 0)
+		seltag = numtags - (-seltag) % numtags;
+	else
+		seltag %= numtags;
+
+	return 1 << seltag;
+}
+
+void
+cycle(const Arg *arg) {
+	const Arg a = { .i = shifttag(arg->i) };
+	view(&a);
+}
+
+void
+tagcycle(const Arg *arg) {
+	const Arg a = { .i = shifttag(arg->i) };
+	tag(&a);
+	view(&a);
+}
+
+int
 main(int argc, char *argv[]) {
 	if(argc == 2 && !strcmp("-v", argv[1]))
 		die("dwm-"VERSION", © 2006-2012 dwm engineers, see LICENSE for details\n");
